File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/application/common/query_params/Pagination.java
package com.c3equalz.user_service.application.common.query_params;

import com.c3equalz.user_service.application.errors.pagination.LimitCantBeNegativeError;
import com.c3equalz.user_service.application.errors.pagination.OffsetCantBeNegativeError;
import lombok.Getter;

@Getter
public class Pagination {
    private final int offset;
    private final int limit;

    public Pagination(int offset, int limit) {
        if (offset < 0) {
            throw new OffsetCantBeNegativeError("Offset cant be negative, got " + offset);
        }

        if (limit < 0) {
            throw new LimitCantBeNegativeError("Limit cant be negative, got " + limit);
        }

        this.offset = offset;
        this.limit = limit;
    }

}

File: сross_platform_software_systems/14/chat_service/src/chat_service/application/common/query_params/pagination.py
from dataclasses import dataclass

from chat_service.application.errors.query_params import PaginationError


@dataclass(frozen=True, slots=True, kw_only=True)
class Pagination:
    offset: int | None = None
    limit: int | None = None

    def __post_init__(self) -> None:
        msg: str | None

        if self.offset is not None and self.offset < 0:
            msg = f"Limit must be greater than 0, got {self.limit}"
            raise PaginationError(msg)

        if self.limit is not None and self.limit <= 0:
            msg = f"Offset must be non-negative, got {self.offset}"
            raise PaginationError(msg)
